Wi-Fi autoconnect and title bar status

tannewt opened the pull request:
This adds support for CIRCUITPY_WIFI_SSID and CIRCUITPY_WIFI_PASSWORD
in `/.env`. When both are defined, CircuitPython will attempt to
connect to the network even when user code isn't running. If the
user code attempts to a network with the same SSID, it will return
immediately. Connecting to another SSID will disconnect from the
auto-connected network. If the user code initiates the connection,
then it will be shutdown after user code exits. (Should match <8
behavior.)

This PR also reworks the default displayio terminal. It now supports
a title bar TileGrid in addition to the (newly renamed) scroll area.
The default title bar is the top row of the display and is positioned
to the right of the Blinka logo when it is enabled. The scroll area
is now below the Blinka logo.

The Wi-Fi auto-connect code now uses the title bar to show its
state including the IP address when connected. It does this through
the "standard" OSC control sequence `ESC ] 0 ; <s> ESC \` where \<s\>
is the title bar string. This is commonly supported by terminals
so it should work over USB and UART as well.

Related to #6174

dhalbert reviewed (changes requested):
I tested with a Feather ESP32-S2 no PSRAM, and also with a FunHouse.

With both, I saw the title bar of the terminal change (neat!) when I connected successfully.

If I gave SSID that could not be reached in `.env`, things failed fairly quickly. However, if I gave an incorrect password in `.env`, there was a long time waiting before the FunHouse showed the status code, and even longer for the /dev/ttyACMx to come up. Sometimes it did not respond at all and I had to disconnect and reconnect (with tio).

I tried going into safe modeso I could edit `.env` by clicking during the three flashes, but it didn't seem to work. It also didn't work on an older build, so this seems like an existing problem. Eventually, it did come up, even with the bad password, and I was able to edit.

So try these scenarios yourself, and see how it works for you. I'm not sure about having the connect be blocking, especially when the connect is not happening, but not blocking may have its own problems.

dhalbert commented on ports/espressif/common-hal/wifi/Radio.c:
```diff
@@ -245,16 +246,31 @@ wifi_radio_error_t common_hal_wifi_radio_connect(wifi_radio_obj_t *self, uint8_t
+    bool connected = ((bits & WIFI_CONNECTED_BIT) != 0) &&
+        !((bits & WIFI_DISCONNECTED_BIT) != 0);
+    if (connected) {
+        if (memcmp(ssid, config->sta.ssid, ssid_len) == 0) {
```
Yeah, it seems to check the password length but not the SSID.

tannewt commented:
> I'm not sure about having the connect be blocking, especially when the connect is not happening, but not blocking may have its own problems.

I've added basic timeout support to `connect()` now. The failure is still kind of long because connect does a full scan first. It won't retry now though so it should be somewhat faster than before. Before it'd retry 5 times before failing.

I've added a TODO to do our own scan first but didn't give in to doing it now.

tannewt commented:
@dhalbert This is ready for another look

dhalbert reviewed (approved):
Thanks for the improved timeout. With a bad password, I am seeing "authentication failure" after about 7 seconds, and CIRCUITPY appears after about 15 seconds. Eventually I see "off" instead of "authentication failure", as I expect you wanted.

Also thanks for the nRF52833 cleanup!
